APPSERV-149 APPSERV-150 APPSERV-151 APPSERV-153 Integrated certificate management documentation (#10)
payara/Payara-Enterprise-Documentation: docs/modules/ROOT/pages/documentation/payara-server/server-configuration/certificate-management.adoc

the DAS), since any certificates added to the instance stores would be lost upon next synchronisation.

*Note*::
This will overwrite an entry already present with the same alias.

[[command-options]]
=== Command Options

[cols="3,1,5,1,1",options="header"]
|===
|Option
|Type
|Description
|Default
|Mandatory

|`--target`
|String
|The name of the instance to add the certificate to.
|server
|no

|`--listener`
|String
|The name of the HTTP or IIOP listener to add the certificate to.
|N/A
|no

|`--domain_name`
|String
|The name of the domain where the target instance exists.
|domain1
|no

|`--domaindir`
|String
|The path to the directory containing the target domain.
|${installDir}/glassfish/domains
|no

|`--file`
|File
|The bundle file to add to the target key store
|N/A
|yes

|`--alias` (primary)
|String
|The alias name to store the certificate bundle in the key store under.
|N/A
|yes

|===

[[example]]
=== Example

[source, shell]
----
asadmin> add-to-keystore --file /home/anon/Downloads/mycert.p12 mycert
----

[[add-to-truststore]]
== `add-to-truststore`

*Usage*::
`asadmin> add-to-truststore --target=instancename --listener=listenername --file /path/to/file alias`

*Aim*::

> This command adds a certificate (e.g. .cert file) to the target instance or listener's trust store.

Reviewer: .cert should be in code format: `.cert`

Reviewer 2: @fturizo, you can simply add a suggestion using the +- icon above the comment input area. It would look like this:
```suggestion
This command adds a certificate (e.g. `.cert` file) to the target instance or listener's trust store.
```

[[command-options]]
=== Command Options

[cols="3,1,5,1,1",options="header"]
|===
|Option
|Type
|Description
|Default
|Mandatory

|`--target`
|String
|The name of the instance to get the certificate from.
|server
|no

|`--listener`
|String
|The name of the HTTP or IIOP listener to get the certificate from.
|N/A
|no

|`--domain_name`
|String
|The name of the domain where the target instance exists.
|domain1
|no

|`--domaindir`
|String
|The path to the directory containing the target domain.
|${installDir}/glassfish/domains
|no

|`--alias` (primary)
|String
|The alias name of the certificate to generate a CSR for.
|N/A
|yes

|===

[[example]]
=== Example

[source, shell]
----
asadmin> generate-csr --listener http-listener-2 --target Instance1 test_cert
----

[[add-to-keystore]]
== `add-to-keystore`

*Usage*::
`asadmin> add-to-keystore --target=instancename --listener=listenername --file /path/to/file alias`

*Aim*::
This command adds a certificate bundle (e.g. .p12 or .jks file) or certificate (e.g. .cert file) to the target instance
or listener's key store using the provided alias.

> If the instance or listener is configured to use the default key store, the command will instead synchronise
> the instance with the DAS (under the assumption the certificate has been added to the default key store of
> the DAS), since any certificates added to the instance stores would be lost upon next synchronisation.

Reviewer 2: Again the same simplification I suggested above:
```suggestion
If the instance or listener is configured to use the default key store, the command will place the certificate bundle in the default key store of the DAS and then synchronise
the instance with the DAS. This is to preserve the certificates across instance restarts because any certificates added directly to the instance stores would be lost upon next synchronisation.
```

Author: See comments elsewhere, your suggestion isn't what happens.

*Note*::
This will overwrite an entry already present with the same alias.

[[command-options]]
=== Command Options

[cols="3,1,5,1,1",options="header"]
|===
|Option
|Type
|Description
|Default
|Mandatory

|`--target`
|String
|The name of the instance to add the certificate to.
|server
|no

|`--listener`
|String
|The name of the HTTP or IIOP listener to add the certificate to.
|N/A
|no

|`--domain_name`
|String
|The name of the domain where the target instance exists.
|domain1
|no

|`--domaindir`
|String
|The path to the directory containing the target domain.
|${installDir}/glassfish/domains
|no

|`--file`
|File

> |The certificate or bundle file to add to the target key store

Reviewer 2: ```suggestion
|The bundle file to add to the target key store
```

|N/A
|yes

|`--alias` (primary)
|String

> |The alias name to store the certificate or bundle the key store under.

Reviewer 2: ```suggestion
|The alias name to store the certificate bundle in the key store under.
```

|N/A
|yes

|===

[[example]]
=== Example

[source, shell]
----
asadmin> add-to-keystore --file /home/anon/Downloads/mycert.p12 mycert
----

[[add-to-truststore]]
== `add-to-truststore`

*Usage*::
`asadmin> add-to-truststore --target=instancename --listener=listenername --file /path/to/file alias`

*Aim*::

> This command adds a certificate bundle (e.g. .p12 or .jks file) or certificate (e.g. .cert file) to the target instance

Reviewer 2: ```suggestion
This command adds a certificate (e.g. .cert file) to the target instance or listener's trust store.
```
Bundles should be added into the keystore using the add-to-keystore command instead.

or listener's trust store using the provided alias.

> If the instance or listener is configured to use the default trust store, the command will instead synchronise
> the instance with the DAS (under the assumption the certificate has been added to the default trust store of
> the DAS), since any certificates added to the instance stores would be lost upon next synchronisation.

Reviewer 2: ```suggestion
If the instance or listener is configured to use the default trust store, the command will place the certificate in the default trust store of the DAS and then synchronise
the instance with the DAS. This is to preserve the certificates across instance restarts because any certificates added directly to the instance stores would be lost upon next synchronisation.
```

Author: See comments elsewhere, your suggestion isn't what happens.

*Note*::
This will overwrite an entry already present with the same alias.

[[command-options]]
=== Command Options

[cols="3,1,5,1,1",options="header"]
|===
|Option
|Type
|Description
|Default
|Mandatory

|`--target`
|String
|The name of the instance to add the certificate to.
|server
|no

|`--listener`
|String
|The name of the HTTP or IIOP listener to add the certificate to.
|N/A
|no

|`--domain_name`
|String
|The name of the domain where the target instance exists.
|domain1
|no

|`--domaindir`
|String
|The path to the directory containing the target domain.
|${installDir}/glassfish/domains
|no

|`--file`
|File

> |The certificate or bundle file to add to the target trust store

Reviewer 2: ```suggestion
|The certificate file to add to the target trust store
```

|N/A
|yes

|`--alias` (primary)
|String

> |The alias name to store the certificate or bundle the trust store under.

Reviewer 2: ```suggestion
|The alias name to store the certificate in the trust store under.
```

|N/A
|yes

|===

[[example]]
=== Example

[source, shell]
----
asadmin> add-to-keystore --file /home/anon/Downloads/mycert.p12 mycert
----

[[remove-from-keystore]]
== `remove-from-keystore`

*Usage*::
`asadmin> remove-from-keystore --target=instancename --listener=listenername alias`

*Aim*::
This command removes a certificate from the target instance or listener's key store matching the provided alias.

> If the instance or listener is configured to use the default key store, the command will instead synchronise
> the instance with the DAS (under the assumption the certificate has been removed from the default key store of
> the DAS), since any certificates removed from the instance stores would be lost upon next synchronisation.

Reviewer 2: ```suggestion
If the instance or listener is configured to use the default key store, the command will remove the certificate from the default key store of the DAS and then synchronise
the instance with the DAS. This is to preserve the certificate changes across instance restarts because any certificates removed directly from the instance stores would be added back upon next synchronisation.
```

Author: See comments elsewhere, your suggestion isn't what happens.

[[command-options]]
=== Command Options

[cols="3,1,5,1,1",options="header"]
|===
|Option
|Type
|Description
|Default
|Mandatory

|`--target`
|String
|The name of the instance to remove the certificate from.
|server
|no

|`--listener`
|String
|The name of the HTTP or IIOP listener to remove the certificate from.
|N/A
|no

|`--domain_name`
|String
|The name of the domain where the target instance exists.
|domain1
|no

|`--domaindir`
|String
|The path to the directory containing the target domain.
|${installDir}/glassfish/domains
|no

|`--alias` (primary)
|String
|The alias name of the certificate to remove.
|N/A
|yes

|===

[[example]]
=== Example

[source, shell]
----
asadmin> remove-from-keystore --domain_name production --target Instance1 --listener http-listener-2 mycert
----

[[remove-from-truststore]]
== `remove-from-truststore`

*Usage*::
`asadmin> remove-from-truststore --target=instancename --listener=listenername alias`

*Aim*::
This command removes a certificate from the target instance or listener's trust store matching the provided alias.

> If the instance or listener is configured to use the default trust store, the command will instead synchronise
> the instance with the DAS (under the assumption the certificate has been removed from the default trust store of
> the DAS), since any certificates removed from the instance stores would be lost upon next synchronisation.

Reviewer 2: ```suggestion
If the instance or listener is configured to use the default trust store, the command will remove the certificate from the default trust store of the DAS and then synchronise
the instance with the DAS. This is to preserve the certificate changes across instance restarts because any certificates removed directly from the instance stores would be added back upon next synchronisation.
```

Author: That isn't what happens.
It doesn't remove it and then sync - syncs are **one-way**. The DAS will always take priority over what has changed on an instance

Author: Given the DAS may not be on the same machine, it doesn't attempt to remove it - it just syncs and warns you

Author: `any certificates removed from the instance stores would be lost upon next synchronisation.`

[[command-options]]
=== Command Options

[cols="3,1,5,1,1",options="header"]
|===
|Option
|Type
|Description
|Default
|Mandatory

|`--target`
|String
|The name of the instance to remove the certificate from.
|server
|no

|`--listener`
|String
|The name of the HTTP or IIOP listener to remove the certificate from.
|N/A
|no

|`--domain_name`
|String
|The name of the domain where the target instance exists.
|domain1
|no

|`--domaindir`
|String
|The path to the directory containing the target domain.
|${installDir}/glassfish/domains
|no

|`--alias` (primary)
|String
|The alias name of the certificate to remove.
|N/A
|yes

|===

[[example]]
=== Example

[source, shell]
----
asadmin> remove-from-truststore --domain_name production --target Instance1 --listener http-listener-2 mycert
----

[[remove-expired-certificates]]
== `remove-expired-certificates`

*Usage*::
`asadmin> remove-expired-certificates --target=instancename --listener=listenername`

*Aim*::
This command removes all expired certificates from the target instance or listener's key and trust stores.

> If the instance or listener is configured to use the default trust store, the command will instead synchronise
> the instance with the DAS (under the assumption the certificate has been removed from the default trust store of
> the DAS), since any certificates removed from the instance stores would be lost upon next synchronisation.

Reviewer 2: ```suggestion
If the instance or listener is configured to use the default key or trust store, the command will remove the certificates from the default key and trust store of the DAS and then synchronise
the instance with the DAS. This is to preserve the certificate changes across instance restarts because any certificates removed directly from the instance stores would be added back upon next synchronisation.
```

Author: See comments above, your suggestion isn't what happens.

[[certificate-management]]
= Integrated SSL/TLS Certificate Management

_Since Payara Server 5.20.0_

A set of commands to provide integrated management of SSL & TLS certificates in Payara Server.

[[generate-self-signed-certificate]]
== `generate-self-signed-certificate`

*Usage*::
`asadmin> generate-self-signed-certificate --target=instancename --listener=listenername --dn="DN1" --alternativenames="ALT1;ALT2;ALT3" alias`

*Aim*::
This command can generate a self-signed certificate for an instance, placing the resultant key pair in the target
instance or listener's key and trust stores.

If the instance or listener is configured to use the default key and trust store, the command will instead synchronise
the instance with the DAS (under the assumption the certificate has been added to the default key and trust store of

> the DAS), since any certificates added to the instance stores would be lost upon next synchronisation.

Author: See comments elsewhere, your suggestion isn't what happens.

*Note*::
While the command will not generate a certificate if one already exists in the target key store under the same alias,

> it will simply overwrite an entry using the same alias in the trust store.

Reviewer 3: This sounds confusing to me. I'd read this as nothing changes but technically the entry gets rewritten as exactly the entry it as before. If so why does the docs have to state this detail?
